hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -753,6 +753,155 @@ struct ConstantFoldingVisitor : OptimizerVisitor, visitor::PreOrder<bool, Consta
     }
 };
 
+// This visitor collects requirement attributes in the AST and toggles unused features.
+struct FeatureRequirementsVisitor : visitor::PreOrder<void, FeatureRequirementsVisitor> {
+    // Lookup table for feature name -> required.
+    using Features = std::map<std::string, bool>;
+
+    // Lookup table for typename -> features.
+    std::map<ID, Features> _features;
+
+    enum class Stage { COLLECT, TRANSFORM };
+    Stage _stage = Stage::COLLECT;
+
+    void collect(Node& node) {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    void transform(Node& node) {
+        _stage = Stage::TRANSFORM;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    void operator()(const declaration::GlobalVariable& x, position_t p) {
+        const auto& id = x.id();
+
+        // We only work on feature flags named `__feat*`.
+        if ( ! util::startsWith(id, "__feat") )
+            return;
+
+        const auto& tokens = util::split(id, "%");
+        assert(tokens.size() == 3);
+
+        // The type name is encoded in the variable name with `::` replaced by `__`.
+        const auto& typeID = ID(util::replace(tokens[1], "__", "::"));
+        const auto& feature = tokens[2];
```

Comment:
I'm wondering if we, instead of using globals for the feature flags, we should introduce a new HILTI-side declaration `%feature ...` that makes it all more explicit. That would then also simplify the code here  to not have to parse things out the ID name (the feature declaration could store the information as part of its AST node).

Reply:
Let's leave this for later (again).